### PR TITLE
(feat) O3-5255: Add optional discharge note to ward patient discharge workspace

### PR DESCRIPTION
**Pull request by sourav-jyoti**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work is based on designs, which are linked or shown either in the Jira ticket or the description below.
- [x] My work includes tests or is validated by existing tests.

## Summary

Adds an optional free-text note field to the ward patient discharge workspace, allowing clinicians to record a reason or context when discharging a patient. The note is saved as an observation using the `consultFreeTextCommentsConcept` on the exit-from-inpatient encounter.

Other changes:
- Rewrites the discharge handler from promise chains to async/await
- Adds Cancel and Confirm buttons (replacing the single ghost button)
- Wires up `hasUnsavedChanges` so closing the workspace with a draft note prompts for confirmation before discarding
- Improves error handling with a structured fallback chain
- Adds test coverage for the discharge workspace

### Before

https://github.com/user-attachments/assets/adb8da7f-a9d3-4cf8-b8ca-3a45a4a8ab06

### After

https://github.com/user-attachments/assets/8246c1af-2644-4976-9ae3-94f019bc0d73

## Related Issue

https://openmrs.atlassian.net/issues/O3-5255

## Other

### CODE REVIEW

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 1]
```diff
@@ -1,13 +1,35 @@
...
+  Workspace2,
+} from '@openmrs/esm-framework';
+import { Controller, useForm } from 'react-hook-form';
+import { type ObsPayload, type WardPatientWorkspaceDefinition, type WardViewContext } from '../../types';
 import { removePatientFromBed, useCreateEncounter } from '../../ward.resource';
 import WardPatientWorkspaceBanner from '../patient-banner/patient-banner.component';
 import styles from './patient-discharge.scss';
 
+const createNoteSchema = (translate: (key: string, defaultMessage: string) => string) =>
+  z.object({
```
[the remark this answers is not in the document]
Comment (sourav-jyoti) [reply packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx]: i have applied this check currently on the frontend as mentioned by anish_dhodari in the [ticket ](https://openmrs.atlassian.net/issues/?filter=-1&selectedIssue=O3-5255) " A **mandatory discharge** note field to record the reason for discharge."

so should i remove it completely

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 2]
```diff
@@ -18,41 +40,80 @@ export default function PatientDischargeWorkspace({
...
+        .then((response) => {
+          if (response?.ok) {
+            if (wardPatient?.bed?.id) {
+              return removePatientFromBed(wardPatient.bed.id, wardPatient?.patient?.uuid);
+            }
+            return response;
+          }
+        })
+        .then((response) => {
+          if (!response?.ok) {
```
Comment (chibongho) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx]: I know the old code had this `response.ok` check, but I don't think we need it.

https://o3-docs.openmrs.org/docs/coding-conventions/error-handling.en-US

**Comment by sourav-jyoti** [comment]
@denniskigen i have updated the according to the new requirements 
1. removed the confirmation modal
2. made the note optional
3. used modern try catch block as suggested by chibongho

### Updated screenrecording

https://github.com/user-attachments/assets/9378abf4-c324-4b0a-8331-be0221594cfc

**Review by denniskigen — COMMENTED** [review]
Thanks for iterating on this, @sourav-jyoti — the async/await conversion is cleaner, and the test coverage is thorough. A few suggestions below.

One thing to update before merge: the PR description still references the confirmation modal and mandatory note field, but the current code has neither (per earlier reviewer feedback). Could you update the description to match the current implementation?

**Comment by sourav-jyoti** [comment]
@denniskigen could you pls review it

**Comment by anish-mekom** [comment]
Hi @denniskigen, requesting for your review on the updates.

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 3]
```diff
@@ -66,10 +89,10 @@ export default function PatientDischargeWorkspace({
         <div className={styles.patientWorkspaceBanner}>
```
Comment (denniskigen) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:89]: This workspace now has editable state (`dischargeNote`), but `Workspace2` doesn't receive `hasUnsavedChanges`. If a user types a note and closes the workspace, the draft is silently lost. Other ward workspaces wire this up, e.g. `cancel-admission-request.component.tsx` and `notes.workspace.tsx`. Please pass `hasUnsavedChanges={dischargeNote.length > 0}`.

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 4]
```diff
@@ -14,47 +20,64 @@ export default function PatientDischargeWorkspace({
...
+          await removePatientFromBed(wardPatient.bed.id, wardPatient?.patient?.uuid);
         }
-      })
-      .catch((err) => {
-        // TODO: better way to handle / display error messages
-        // https://openmrs.atlassian.net/browse/O3-5423
+
+        showSnackbar({
+          title: t('patientWasDischarged', 'Patient was discharged'),
+          kind: 'success',
```
Comment (denniskigen) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:57]: In the original code, `mutate()` was called in `finally`, so it ran on both success and failure. Moving it into the `try` block means a partial failure (encounter created but bed removal throws) leaves the ward view stale. Consider moving it back to `finally`.

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 5]
```diff
@@ -83,20 +106,37 @@ export default function PatientDischargeWorkspace({
...
+            <ResponsiveWrapper>
+              <TextArea
+                value={dischargeNote}
+                onChange={(e) => setDischargeNote(e.target.value)}
+                labelText={t('dischargeNote', 'Discharge note (optional)')}
+                placeholder={t('dischargeNotePlaceholder', 'Write any notes here')}
+              />
+            </ResponsiveWrapper>
+          </div>
           <ButtonSet className={styles.buttonSet}>
```
Comment (denniskigen) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:122]: Two things: "discharging" should be sentence case ("Discharging"), and concatenating the ellipsis is fragile for i18n. Consider `t('discharging', 'Discharging...')` as a single translatable string.

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.scss [hunk 6]
```diff
@@ -14,28 +13,18 @@
   margin: (-(layout.$spacing-05)) (-(layout.$spacing-05)) layout.$spacing-05 (-(layout.$spacing-05));
 }
 
-.contentSwitcher {
-  margin-top: layout.$spacing-03;
-}
-
 .workspaceForm {
   margin-top: layout.$spacing-05;
   flex: 1;
```
Comment (denniskigen) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.scss:18]: Nit: double parentheses `((layout.$spacing-05))` — the outer pair is unnecessary.

**Review by denniskigen — DISMISSED** [review]
Thanks for the iteration here, @sourav-jyoti. The async/await rewrite and test coverage are both solid. A few things to address before this is ready.

**Review by jayasanka-sack — DISMISSED** [review]
@sourav-jyoti the E2E test and build needs fixing. can you take a look and let us know?

**Comment by sourav-jyoti** [comment]
@jayasanka-sack done

**Comment by anish-mekom** [comment]
Hi @jayasanka-sack requesting your review.

File: packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx [hunk 7]
```diff
@@ -14,62 +20,79 @@ export default function PatientDischargeWorkspace({
...
-          return response;
-        }
-      })
-      .then((response) => {
-        if (response?.ok) {
-          showSnackbar({
-            title: t('patientWasDischarged', 'Patient was discharged'),
-            kind: 'success',
+      try {
+        const note = dischargeNote;
```
Comment (denniskigen) [inline packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx]: Small polish suggestion: since the discharge note is optional, could we trim it before building `obs`? Otherwise whitespace-only input (e.g. `"   "`) creates a blank note observation instead of behaving like no note.

For example:

```tsx
const note = dischargeNote.trim();
```

It would also be worth adding a small test that whitespace-only input submits with an empty `obs` array.

Comment (sourav-jyoti) [reply packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx]: @denniskigen added .trim()
and also migrated jest to vi

Comment (sourav-jyoti) [reply packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx]: @denniskigen @jayasanka-sack requesting ur review

**Review by chibongho — APPROVED** [review]
I discussed this with my team internally. We are ok with this change. (If we need to turn off the confirmation dialog on discharge we'll add a config later.) Thanks!

**Review by denniskigen — APPROVED** [review]
Thanks for seeing this through, @sourav-jyoti — really solid work. The async/await rewrite, error fallback chain, and test coverage are all well done. I pushed one small fix on top:  now uses  to stay consistent with the trim logic in , so whitespace-only input no longer triggers the unsaved-changes prompt.

**Review by denniskigen — APPROVED** [review]
Thanks for seeing this through, @sourav-jyoti. Really solid work. The async/await rewrite, error fallback chain, and test coverage are all well done. I pushed one small fix on top: `hasUnsavedChanges` now uses `dischargeNote.trim().length > 0` to stay consistent with the trim logic in `onSubmit`, so whitespace-only input no longer triggers the unsaved-changes prompt.
